### DISCUSSION TITLE
Container reports "Error: failed to find BLE beacon"  when deploying keys

### DISCUSSION THREAD

**Discussion by abreingan** [tesla-local-control/tesla_ble_mqtt_docker#8] [Q&A] [answered]
Hi all, I suspect this is a newbie docker / BLE question, and while I have tried a few of the options google throws up I haven't got anything working. Any tips gratefully received.

All goes well till the Deploy Key stage which I guess is the first use of BLE. It fails with:- 
tesla_ble_mqtt  | Deploying public key to vehicle
tesla_ble_mqtt  | Error: failed to find BLE beacon for LRWYHCEK7PC906935 (S893f531b20c5ac79C): can't scan: context deadline exceeded

However in a terminal the system can find the car, as bluetoothctl and scan on does show the device which the container says it is looking for. 

I am blundering around here a bit, as I haven't much docker experience, but I can confirm 
the hardware is a Raspberry Pi 3B, default latest lite os. I tried that first and when that didn't work I tried installing  dbus-broker, and running 
systemctl enable dbus-broker.service
systemctl --global enable dbus-broker.service

That made no difference so I cast myself on your mercy rather than digging deeper (:-)>

**Comment by iainbullock** [comment]
Possibly your bluetooth device is too far away, or more likely the car is asleep. Try waking the car using the Tesla App

**Reply by abreingan** [reply]
Thanks for that Ian, but I think I had that covered. The Pi was about 2m from the front of the car and it was definitely awake. I will keep troubleshooting.

**Reply by abreingan** [reply]
I am trying to troubleshoot why bluetooth seems to work in the system but not in the container I ran the image to see what bluetoothctl did. So issuing:- 
docker run --net=host --privileged -i -t iainbullock/tesla_ble_mqtt
  and then
bluetoothctl
I get the error:- 
Waiting to connect to bluetoothd...dbus[8]: arguments to dbus_connection_get_object_path_data() were incorrect, assertion "connection != NULL" failed in file dbus-connection.c line 5969.
This is normally a bug in some application using the D-Bus library.
D-Bus not compiled with backtrace support so unable to print a backtrace

Now obviously this is working for others so it may be a configuration on my part. I stress I am new at docker but hoping it might give someone a clue of what I have done wrong

**Comment by iainbullock** [comment]
If you are not already doing so use docker-compose.yml when starting the container as described here https://github.com/iainbullock/tesla_ble_mqtt_docker/blob/main/INSTALL.md

I think you are not providing the mapping between /run/dbus from host to container if you start the container using docker run --net=host --privileged -i -t iainbullock/tesla_ble_mqtt

Hope that helps!

**Reply by abreingan** [reply]
Sorry Iain, that's me blundering around - its my first attempt to spin up a docker container and I am reading the manual as I go. I will describe what I did below hoping someone will point out what I missed, but the upshot is that I get the same error
tesla_ble_mqtt  | Deploying public key to vehicle
tesla_ble_mqtt  | Error: failed to find BLE beacon for LRWYHCEK7PC906935 (S893f531b20c5ac79C): can't scan: context deadline exceeded
I have run this multiple times with the pi in different positions around the car, and the last time with the door open actually just inside the car, so I don't think it is a range issue. The car was always awake.

I have tried a bluetoothctl scan both in a system terminal and (I think) in the container by running 
docker exec -it 37f659151f28 bluetoothctl in another terminal. both show the same thing - two Shelly switches that are nearby and the car - the log is below. 

So I am clearly missing something but I cannot think of any further troubleshoot options. 

for clarity (and because I probably missed something)
I went back and rebuild the machine from scratch a Raspberry Pi 3b with raspios-bookworm-arm64-lite.
sudo apt-get update then sudo apt-get upgrade
set up network etc 
Install docker 
docker pull iainbullock/tesla_ble_mqtt
create a project directory and copy in the amended docker-compose.yml
docker  volume create tesla_ble_mqtt 
docker compose up - that shows the connection to Mqtt and I can then Generate Keys sucessfully. however the deploy key always returns
tesla_ble_mqtt  | Configuration deploy_key requested
tesla_ble_mqtt  | Deploying public key to vehicle
tesla_ble_mqtt  | Error: failed to find BLE beacon for LRWYHCEK7PC906935 (S893f531b20c5ac79C): can't scan: context deadline exceeded

running 
docker exec -it 37f659151f28 bluetoothctl in another terminal then "scan on"  seems to show the beacon that could not be found - the log is below:-
alb@raspble:~ $ docker exec -it 37f659151f28 bluetoothctl
hci0 new_settings: powered bondable ssp br/edr le secure-conn
Agent registered
[CHG] Controller B8:27:EB:17:57:8C Pairable: yes
[bluetooth]# scan on
SetDiscoveryFilter success
hci0 type 7 discovering on
Discovery started
[CHG] Controller B8:27:EB:17:57:8C Discovering: yes
[NEW] Device A8:03:2A:B6:CE:C6 ShellyPlus1-A8032AB6CEC4
[NEW] Device 44:17:93:AA:04:6E ShellyPlus1-441793AA046C
[NEW] Device 74:46:B3:AE:E6:F2 S893f531b20c5ac79C
[CHG] Device 44:17:93:AA:04:6E RSSI: 0xffffffcb (-53)

**Reply by iainbullock** [reply]
You seem to have done everything right here but if the car BLE beacon never shows when you do a bluetoothctl scan on then it suggests a hardware issue. 

Do you have any other devices you can try it with?

Another thought are you connecting to your RPi using 2.4GHz WiFi. I have seen some reports that this can interfere with the Bluetooth. If supported connect via 5GHz WiFi or better via Ethernet

**Reply by abreingan** [reply]
Sorry I obviously didnt explain clearly - I think the car BLE beacon does show on a bluetoothctl scan, even in the container. it is the 4th line down in the scan (74:46:B3:AE:E6:F2 S893f531b20c5ac79C) which is the same as the error (Error: failed to find BLE beacon for LRWYHCEK7PC906935 (S893f531b20c5ac79C):)
That's why I am so confused.
[CHG] Controller B8:27:EB:17:57:8C Discovering: yes
[NEW] Device A8:03:2A:B6:CE:C6 ShellyPlus1-A8032AB6CEC4
[NEW] Device 44:17:93:AA:04:6E ShellyPlus1-441793AA046C
[NEW] Device 74:46:B3:AE:E6:F2 S893f531b20c5ac79C
[CHG] Device 44:17:93:AA:04:6E RSSI: 0xffffffcb (-53)

The pi is connected to the network by ethernet (I am old fashioned) so the radio is only used for bluetooth. On my first attempt I turned wifi off in config, but forgot when I rebuilt it so I have tried both ways!

**Reply by iainbullock** [reply, marked as answer]
Ah yes I see what you mean it is there. 

Things to try: 

- Not using WiFi for the RPi if you currently are

- Try different locations for the RPi

- Try a different device if you have one

- Try to deploy the key multiple times in succession (you probably already have)

**Reply by abreingan** [reply]
Thanks Iain, I will try and get my hands on a different device

**Reply by abreingan** [reply]
OK, borrowed a Pi 4 and everything worked perfectly. So it was a hardware issue.
Thank you very much for your help (and patience).
